commit/merkleroot/rmn/peerclient.go

```diff
@@ -31,27 +52,150 @@ type PeerResponse struct {
+	mu                          *sync.RWMutex
 }
 
-func NewPeerClient(peer ragep2ptypes.PeerInfo, rmnNodeIDs []rmntypes.NodeID) PeerClient {
+func NewPeerClient(
+	lggr logger.Logger,
+	peerGroupFactory networking.PeerGroupFactory,
+	bootstrappers []commontypes.BootstrapperLocator,
+) PeerClient {
 	return &peerClient{
-		peer:       peer,
-		rmnPeerIDs: mapset.NewSet(rmnNodeIDs...),
-		respChan:   make(chan PeerResponse),
+		lggr:                        lggr,
+		peerGroupFactory:            peerGroupFactory,
+		respChan:                    make(chan PeerResponse),
+		peerGroup:                   nil,
+		rageP2PStreams:              make(map[rmntypes.NodeID]networking.Stream),
+		genericEndpointConfigDigest: cciptypes.Bytes32{},
+		bootstrappers:               bootstrappers,
+		mu:                          &sync.RWMutex{},
+	}
+}
+
+func (r *peerClient) InitConnection(
+	_ context.Context,
+	commitConfigDigest, rmnHomeConfigDigest cciptypes.Bytes32,
+	peerIDs []string,
+) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	h := sha256.Sum256(append(commitConfigDigest[:], rmnHomeConfigDigest[:]...))
+	r.genericEndpointConfigDigest = writePrefix(ocr2types.ConfigDigestPrefixCCIPMultiRoleRMNCombo, h)
+
+	peerGroup, err := r.peerGroupFactory.NewPeerGroup(
+		[32]byte(r.genericEndpointConfigDigest),
+		peerIDs,
+		r.bootstrappers,
+	)
+
+	if err != nil {
+		return fmt.Errorf("new peer group: %w", err)
+	}
+
+	r.peerGroup = peerGroup
+	return nil
+}
+
+func (r *peerClient) Close() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// individual streams are closed by the peer group
+	if err := r.peerGroup.Close(); err != nil {
+		return fmt.Errorf("close peer group: %w", err)
+	}
+
+	return nil
+}
+
+func (r *peerClient) Send(rmnNode rmntypes.HomeNodeInfo, request []byte) error {
+	if r.peerGroup == nil {
+		return ErrNoConn
+	}
+
+	stream, err := r.getOrCreateRageP2PStream(rmnNode)
+	if err != nil {
+		return fmt.Errorf("get or create rage p2p stream: %w", err)
+	}
+
+	r.mu.Lock()
+	stream.SendMessage(request)
+	r.mu.Unlock()
+
+	return nil
+}
+
+func (r *peerClient) getOrCreateRageP2PStream(rmnNode rmntypes.HomeNodeInfo) (networking.Stream, error) {
+	r.mu.RLock()
+	stream, ok := r.rageP2PStreams[rmnNode.ID]
+	r.mu.RUnlock()
+	if ok {
+		return stream, nil
 	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// todo: versioning for stream names e.g. for 'v1_7'
+	streamName := fmt.Sprintf("ccip-rmn/v1_6/%x", r.genericEndpointConfigDigest)
+	r.lggr.Info("Creating new stream", "streamName", streamName)
+
+	var err error
+	stream, err = r.peerGroup.NewStream(
+		rmnNode.PeerID.String(),
+		networking.NewStreamArgs1{
+			StreamName:         streamName,
+			OutgoingBufferSize: 1,
+			IncomingBufferSize: 1,
+			MaxMessageLength:   2_097_152, // 2MB
+			MessagesLimit: ragep2p.TokenBucketParams{
+				Rate:     20,
+				Capacity: 100,
+			},
+			BytesLimit: ragep2p.TokenBucketParams{
+				Rate:     20_971_520,  // 20MB
+				Capacity: 104_857_600, // 100MB
+			},
```

Comment:
Should these be defined via configuration or are these good enough to be constants?